### PR TITLE
Adjusted font size in trainings constructions

### LINKED ISSUE #3: Adjust font size in training instructions
**As a** user, **I want** to have the training instructions for models in larger font size, **so that**  I can read them more easily.

### Acceptance Criteria

- [x] Adjust the font size dynamically based on the screen size;
- [x] User can increase or decrease the font size;

### Additional Resources:

image

### Notes

Effort 5

### PR DESCRIPTION
**Pull request by RukiyyeE**
### Resolves https://github.com/claassenlab/cellpose/issues/3
----
### Adjusted font size

**The following changes have been made in `TrainHelpWindow` class:**

- layout is now directly assigned to the dialog window: `self.setLayout(layout)`
- label has been made an instance attribute self.label:
   - `self.label = QLabel(text)`
      `self.label.setWordWrap(True)`
      `layout.addWidget(self.label, 0, 0, 1, 1)`

- created dropdown menu to adjust the font size
- added the `adjust_font_size` method so that users can adjust font size
- `resizeEvent` has been overridden to call adjust `font size` when the window is resized, ensuring that the window size adjusts when the window size changes

### CODE REVIEW

File: cellpose/gui/guiparts.py [hunk 1]
```diff
@@ -1,10 +1,18 @@
 """
-Copyright © 2023 Howard Hughes Medical Institute, Authored by Carsen Stringer and Marius Pachitariu.
+Copyright Â© 2023 Howard Hughes Medical Institute, Authored by Carsen Stringer and Marius Pachitariu.
```
Comment (RenzoRichter) [inline cellpose/gui/guiparts.py]: Warum hier Â

File: cellpose/gui/guiparts.py [hunk 2]
```diff
@@ -223,15 +231,10 @@ def __init__(self, parent, model_strings):
         # choose parameters
         labels = ["learning_rate", "weight_decay", "n_epochs", "model_name"]
         self.edits = []
-        self.parameter_explanations = ["The learning rate determines how quickly or slowly the model learns from data. A higher learning rate may lead to faster learning but could cause the model to overshoot the optimal solution. Conversely, a lower learning rate may result in slower learning but is safer and more likely to find the best solution.",
```
Comment (RenzoRichter) [inline cellpose/gui/guiparts.py]: Note to self: Wurde hinzugefügt in PR #9

File: cellpose/gui/guiparts.py [hunk 3]
```diff
@@ -223,15 +231,10 @@ def __init__(self, parent, model_strings):
...
-                                       "Weight decay helps prevent overfitting by penalizing large parameter values in the model. \n Increasing weight decay encourages the model to learn simpler patterns from the data,\n improving its ability to generalize to new, unseen examples.",
-                                       "The number of times the entire dataset is passed forward and backward through the machine learning model during training. Increasing the number of epochs allows the model to see the data more times, potentially improving its accuracy. However, too many epochs can lead to overfitting, where the model memorizes the training data instead of learning generalizable patterns.",
-                     ""]
         yoff += 1
         for i, label in enumerate(labels):
             qlabel = QLabel(label)
             qlabel.setAlignment(QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
-            qlabel.setToolTip(self.parameter_explanations[i])
```
Comment (RenzoRichter) [inline cellpose/gui/guiparts.py]: Note to self: Wurde hinzugefügt in PR #9

File: cellpose/gui/guiparts.py [hunk 4]
```diff
@@ -337,28 +340,54 @@ def __init__(self, parent=None):
...
-        label.setWordWrap(True)
-        layout.addWidget(label, 0, 0, 1, 1)
+        self.label = QLabel(text)
+        self.label.setWordWrap(True)
+        layout.addWidget(self.label, 0, 0, 1, 1)
+
+        # Dropdown menu for font size
+        self.font_size_combo = QComboBox(self)
+        self.font_size_combo.addItems([str(size) for size in range(5, 30, 3)])
+        self.font_size_combo.currentIndexChanged.connect(self.adjust_font_size)
```
Comment (RenzoRichter) [inline cellpose/gui/guiparts.py:372]: Die Line `self.font_size_combo.currentIndexChanged.connect(self.adjust_font_size)` braucht am besten noch ein erklärender Kommentar 
⇒ Sonst ein bisschen unverständlich
z.B.:
```
# Diese Zeile verbindet das Signal "currentIndexChanged" des QComboBox-Objekts "self.font_size_combo" mit der Methode "self.adjust_font_size".
# Das bedeutet, dass jedes Mal, wenn der aktuell ausgewählte Index in "self.font_size_combo" geändert wird (z.B. durch Benutzerinteraktion), 
# die Methode "self.adjust_font_size" automatisch aufgerufen wird.
```
Gerne auch noch ein bisschen gekürzt :)

File: cellpose/gui/guiparts.py [hunk 5]
```diff
@@ -337,28 +340,54 @@ def __init__(self, parent=None):
...
+        font = QFont("Arial", min(font_size, new_font_size))
+        self.label.setFont(font)
+
+    def resizeEvent(self, event):
+        # Call adjust_font_size when the window is resized
+        self.adjust_font_size()
+        super().resizeEvent(event)
+
+
+if __name__ == "__main__":
```
Comment (RenzoRichter) [inline cellpose/gui/guiparts.py]: Was wird hier genau gemacht? 
Kommentar z.B.:
```
# Dieser Code startet die TrainHelpWindow-Anwendung, wenn das Skript direkt ausgeführt wird.
# Das TrainHelpWindow ist ein Dialogfenster, welches es dem Nutzer ermöglicht ???
```

File: cellpose/gui/guiparts.py [hunk 6]
```diff
@@ -337,28 +340,54 @@ def __init__(self, parent=None):
...
-        label.setFont(QtGui.QFont("Arial", 8))
-        label.setWordWrap(True)
-        layout.addWidget(label, 0, 0, 1, 1)
+        self.label = QLabel(text)
+        self.label.setWordWrap(True)
+        layout.addWidget(self.label, 0, 0, 1, 1)
+
+        # Dropdown menu for font size
+        self.font_size_combo = QComboBox(self)
+        self.font_size_combo.addItems([str(size) for size in range(5, 30, 3)])
```
Comment (RenzoRichter) [inline cellpose/gui/guiparts.py]: Menü geht nur bis 29, wenn man in 3er-Schritten von 5 hochgeht.

File: cellpose/gui/guiparts.py [hunk 7]
```diff
@@ -337,28 +340,54 @@ def __init__(self, parent=None):
...
-            "guitrainhelpwindowtext.html")
+        text_file = pathlib.Path(__file__).parent.joinpath("guitrainhelpwindowtext.html")
         with open(str(text_file.resolve()), "r") as f:
             text = f.read()
 
-        label = QLabel(text)
-        label.setFont(QtGui.QFont("Arial", 8))
-        label.setWordWrap(True)
-        layout.addWidget(label, 0, 0, 1, 1)
+        self.label = QLabel(text)
```
Comment (RenzoRichter) [inline cellpose/gui/guiparts.py:355]: Setzt das `label` vielleicht von Beginn an auf eine sinnvolle Schriftgröße z.B. 17
z.B.:
`label.setFont(QtGui.QFont("Arial", 17))`

**Review by RenzoRichter — CHANGES_REQUESTED** [review]
Hey there, great job. 

## MAIN-Review:
 Alle Anforderungen aus dem Issue sind erfüllt (⇒ code implementiert Issue [#3  ])

Code compiles und die Menüs funktionieren. 
Auch mehrere Menüs gleichzeitig offen scheinen gut zu funktionieren.

_______________________________________
## Nitpicks

Fügt vielleicht noch ein paar extra Kommentare ein, um die Lesbarkeit noch weiter zu steigern.

Außerdem setzt bitte die vorgeschlagenen Änderungen noch in die Tat um,
siehe Review-Kommentare.

______________________________________
### File can not yet be merged to main!

File: cellpose/gui/guiparts.py [hunk 8]
```diff
@@ -1,10 +1,18 @@
 """
-Copyright © 2023 Howard Hughes Medical Institute, Authored by Carsen Stringer and Marius Pachitariu.
+Copyright Â© 2023 Howard Hughes Medical Institute, Authored by Carsen Stringer and Marius Pachitariu.
 """
 
+from tkinter import *
+import tkinter as tk
+from tkinter import ttk
```
Comment (jakobjo) [inline cellpose/gui/guiparts.py]: Werden diese tkinter imports genutzt?

File: cellpose/gui/guiparts.py [hunk 9]
```diff
@@ -337,28 +340,54 @@ def __init__(self, parent=None):
...
+
+        self.adjust_font_size()  # Initial font size adjustment
+
         self.show()
 
+    def adjust_font_size(self):
+        # Get the current font size from the combo box
+        font_size = int(self.font_size_combo.currentText())
+        # Calculate the new font size based on window height
+        new_font_size = max(5, int(self.height() / 30))
```
Comment (jakobjo) [inline cellpose/gui/guiparts.py]: Das ist eine gute Umsetzung, echt stark.

Beim skalieren der Schriftgröße sollte allerdings auch die Breite des Fensters berücksichtigt werden, da es sonst bei einem schmalen aber hohen Fenster und großer Schriftgröße selbts adjustziert (und das gerade zu explosionsartig). Ich habe ein bisschen rumprobiert und vermute, dass QT beim resizen immer versucht den gesamten Text des Labels darzustellen und deshalb das Fenter nach unten erweitert. 

Um das zu verhindern, könnte man eventuell die Berechnung der new_font_size anpassen und auch von der Breite des Fensters abhängig machen, da sonst bei großer Schrift die Größe als angemessen erachtet wird obwohl das Fenster zu schmal ist. Andere Idee wäre sonst, da Fenster in einem festen Verhältnis nur anpassbar zu machen (z.B. 16:9), aber der erste Weg wäre auf jedenfall optimaler.

File: cellpose/gui/guiparts.py [hunk 10]
```diff
@@ -337,28 +340,54 @@ def __init__(self, parent=None):
...
-        layout.addWidget(label, 0, 0, 1, 1)
+        self.label = QLabel(text)
+        self.label.setWordWrap(True)
+        layout.addWidget(self.label, 0, 0, 1, 1)
+
+        # Dropdown menu for font size
+        self.font_size_combo = QComboBox(self)
+        self.font_size_combo.addItems([str(size) for size in range(5, 30, 3)])
+        self.font_size_combo.currentIndexChanged.connect(self.adjust_font_size)
+        layout.addWidget(self.font_size_combo, 1, 0, 1, 1)
```
Comment (jakobjo) [inline cellpose/gui/guiparts.py:373]: Die Kombobox ist etwas groß am unteren Fensterrand.

**Review by jakobjo — CHANGES_REQUESTED** [review]
Hi, echt gute Arbeit, die gewünschten Punkte sind implementert und funktionieren fast einwandfrei (s. Kommentar). Ich nehme mal an, dass ihr nicht jedes Leerzeichen selbst gesetzt habt hinter den Hastags der Kommentar, aber solche Veränderungen machen die eigentlichen Änderungen etwas unübersichtlich (Vielleicht wurde das irgendwie automatisch gemacht?).  
Wenn ihr Fragen habt was genau für ein Problem mit dem automatischen Anpassen vorliegt kontaktiert mich gerne. 

Zudem würde ich die default Fenstergröße etwas erhöhen und eine minimale Größe festlegen.

Ansonsten würde ich mich den Punkten von Henry anschließen, vor allem die default Fontsize zu erhöhen.

**Comment by RukiyyeE** [comment]
> Hi, echt gute Arbeit, die gewünschten Punkte sind implementert und funktionieren fast einwandfrei (s. Kommentar). Ich nehme mal an, dass ihr nicht jedes Leerzeichen selbst gesetzt habt hinter den Hastags der Kommentar, aber solche Veränderungen machen die eigentlichen Änderungen etwas unübersichtlich (Vielleicht wurde das irgendwie automatisch gemacht?). Wenn ihr Fragen habt was genau für ein Problem mit dem automatischen Anpassen vorliegt kontaktiert mich gerne.
> 
> Zudem würde ich die default Fenstergröße etwas erhöhen und eine minimale Größe festlegen.
> 
> Ansonsten würde ich mich den Punkten von Henry anschließen, vor allem die default Fontsize zu erhöhen.

Danke für euren Review. Die Leerzeichen habe ich auch erst nach dem pushen bemerkt. Sollen die Leerzeichen Rückgängig gemacht werden oder soll es diesmal so bleiben?

File: cellpose/gui/guiparts.py (hunk 6 above)
Comment (RukiyyeE) [reply cellpose/gui/guiparts.py]: Ist das hier also so in Ordnung?

**Comment by jakobjo** [comment]
> > Hi, echt gute Arbeit, die gewünschten Punkte sind implementert und funktionieren fast einwandfrei (s. Kommentar). Ich nehme mal an, dass ihr nicht jedes Leerzeichen selbst gesetzt habt hinter den Hastags der Kommentar, aber solche Veränderungen machen die eigentlichen Änderungen etwas unübersichtlich (Vielleicht wurde das irgendwie automatisch gemacht?). Wenn ihr Fragen habt was genau für ein Problem mit dem automatischen Anpassen vorliegt kontaktiert mich gerne.
> > Zudem würde ich die default Fenstergröße etwas erhöhen und eine minimale Größe festlegen.
> > Ansonsten würde ich mich den Punkten von Henry anschließen, vor allem die default Fontsize zu erhöhen.
> 
> Danke für euren Review. Die Leerzeichen habe ich auch erst nach dem pushen bemerkt. Sollen die Leerzeichen Rückgängig gemacht werden oder soll es diesmal so bleiben?

Gerne kurz rückgängig machen, dann gibts beim mergen da keine Probleme.

File: cellpose/gui/guiparts.py (hunk 7 above)
Comment (mariehartung) [reply cellpose/gui/guiparts.py:355]: Ich glaube, das ist aber eh das Fenster "Help with GUI", nicht das für uns. Das lässt sich aber soweit ich das sehe sowieso nicht öffnen (unabhängig von unserem Code). Wir erhöhen die default Größe aber auf jeden Fall mal bei uns.

File: cellpose/gui/guiparts.py [hunk 11]
```diff
@@ -370,10 +372,14 @@ def __init__(self, parent=None):
     def adjust_font_size(self):
         # Get the current font size from the combo box
         font_size = int(self.font_size_combo.currentText())
-        # Calculate the new font size based on window height
-        new_font_size = max(5, int(self.height() / 30))
+        # Calculate the new font size based on window height and width
+        new_font_size = max(5, max(int(self.height() / 50), int(self.width() / 60)))
```
Comment (jakobjo) [inline cellpose/gui/guiparts.py]: Das funktioniert leider immer noch nicht, das zuvor beschriebene Problem besteht weiterhin. Ich schicke euch mal ein Video auf Discord um euch zu zeigen was ich meine.

**Review by RenzoRichter — APPROVED** [review]
Hey there, great job.

# MAIN-Review:
Alle Anforderungen aus dem Issue sind erfüllt (⇒ code implementiert Issue [#3  ])

Code compiles und die Menüs funktionieren.
Auch mehrere Menüs gleichzeitig offen scheinen gut zu funktionieren.

---

# Nitpicks
Fügt vielleicht noch ein paar extra Kommentare ein, um die Lesbarkeit noch weiter zu steigern.

---

## File can be merged to main!

**Review by jakobjo — APPROVED** [review]
## Main Review

Das implentierte Feature erfüllt alle gewünschten Bedingungen aus  Issue #3.
Die Startschriftgröße und Fenstergröße wurden auf einen sinnvollen Wert angepasst und der Text passt sich dynamisch mit der Fenstergröße an. Ebenso funktioniert die Schriftgrößenauswahl durch den User selbst.

## Requested Changes

Alle geforderten Veränderungen wurden implementiert, die Fenstergröße lässt sich nun ohne Probleme anpassen und die Startgröße ist gut gewählt.

## Branch can be merged!

Ich weiß nicht genau wie das passiert ist, aber euer Branch wurde auf einem Commit erstellt, der die Änderungen des anderen Scrum Teams bei den Training Prametern bereits enthalten hat. Leider wurden die bei euch wieder verworfen, das wird vermutlich zu einem Konflikt führen, schaut also, dass diese Änderungen erhalten bleiben. (Gemeint ist die Parameter Explanation beim drüberhovern).

Gute Arbeit!

**Review by jakobjo — APPROVED** [review]
Das alte Feature wurde wieder hinzugefügt und funktioniert. Der Branch kann mit dem Mainbranch gemerged werden.

**Review by RenzoRichter — APPROVED** [review]
Feature Text des anderen Features (#5) wurde wieder hinzugefügt. 
Funktionalität besteht weiterhin. 

---
## Der Branch kann auf Main gemerged werden
